gprofiler/spark/sampler.py

```diff
@@ -258,31 +61,31 @@ def _collect_loop(self) -> None:
         ), "A valid API client or storage directory is required"
         timefn = time.monotonic
         start_time = timefn()
+        discovered = False
+
         while not self._stop_event.is_set():
-            if self._spark_sampler is None:
-                spark_cluster_conf = self._find_spark_cluster()
-                if spark_cluster_conf is not None:
-                    master_address, cluster_mode = spark_cluster_conf
-                    self._spark_sampler = SparkCollector(cluster_mode, master_address)
-                else:
-                    if timefn() - start_time >= FIND_CLUSTER_TIMEOUT_SECS:
+            if self._big_data_sampler is not None:
+                if not discovered:
+                    if (
+                        not (discovered := self._big_data_sampler.discover())
+                        and timefn() - start_time >= FIND_CLUSTER_TIMEOUT_SECS
+                    ):
                         logger.info("Timed out identifying Spark cluster. Stopping Spark collector.")
                         break
```

Comment:
Nicer as 2 separate `if`s, move readable:
```suggestion
                    discovered = self._big_data_sampler.discover()

                if not discovered and timefn() - start_time >= FIND_CLUSTER_TIMEOUT_SECS:
                    logger.info("Timed out identifying Spark cluster. Stopping Spark collector.")
                    break
```

Reply:
https://github.com/Granulate/gprofiler/pull/749/files#diff-18dce65a9628cdb97bacc94da9b0401f8686d3326c5a71c91a23eb0bf4cbb919R67-R72